<a href="https://akademie.datamics.com/kursliste/">![title](bg_datamics_top.png)</a>

<center><em>© Datamics</em></center><br><center><em>Besuche uns für mehr Informationen auf <a href='https://akademie.datamics.com/kursliste/'>www.akademie.datamics.com</a></em>

# Stringformatierung

Stringformatierung erlaubt es uns, Elemente in einen String einzufügen, anstatt diese mit Komma und Plus zusammenzufügen. Hier ein schneller Vergleich:


    spieler = 'Thomas'
    punkte = 33
    
    'Letzte Nacht hat ' + spieler + ' insgesamt ' + str(punkte) + ' Punkte gesammelt.'  # Konkatenation
    
    f'Letzte Nacht hat {player} insgesamt {punkte} Punkte gesammelt.'          # Stringformatierung


Es gibt drei Methoden für die Stringformatierung.
* Die älteste Methode verwendet Platzhalter mit dem Modulo-Zeichen `%`
* Angenehmer in der Anwendung ist die String-Methode `.format()`.
* Die neuste, mit Python 3.6 eingeführte Methode, verwendet *f-Strings* genannte, formatierte Stringliterale.

Da du wahrscheinlich bei der Arbeit mit Code von Anderen auf alle drei Methoden treffen wirst, beschreiben wir hier alle drei. 

## Formatierung mit Platzhaltern
Du kannst <code>%s</code> verwenden, um Strings in dein Print-Statement einzufügen. Der Modulo-Operator `%` wird in diesem Zusammenhang als "Stringformatierungsoperator" bezeichnet.

In [1]:
print("Hier werde ich %s einfügen." % 'etwas')

Hier werde ich etwas einfügen.


Mehrere Elemente können mit einem Tupel nach dem `%`-Operator eingefügt werden.

In [2]:
print("Ich werde hier %s Text einfügen und einen %s hier." % ('einen','weiteren'))

Ich werde hier einen Text einfügen und einen weiteren hier.


Du kannst auch Variablennamen übergeben:

In [3]:
x, y = 'einen', 'weiteren'
print("Ich werde hier %s Text einfügen und einen %s hier." % (x, y))

Ich werde hier einen Text einfügen und einen weiteren hier.


### Formatkonvertierungsmethoden
An dieser Stelle ist anzumerken, dass die Operatoren <code>%s</code> und <code>%r</code> ein beliebiges Objekt in einen String konvertieren anhand der Methoden `str()` und `repr()`. Wir werden in späteren Kapiteln mehr über diese beiden Funktionen lernen, du solltest aber beachten, dass `%r` und `repr()` die *Stringrepräsentation* des Objektes zurückgeben, inklusive Hochkommata und Maskierungszeichen.

In [4]:
print('Er sagte, sein Name sei %s.' % 'Fred')
print('Er sagte, sein Name sei %r.' % 'Fred')

Er sagte, sein Name sei Fred.
Er sagte, sein Name sei 'Fred'.


Mit dem maskierten t `\t` wird ein Tabulator eingefügt: 

In [5]:
print('Wir sind %s.' % 'am \tbraten')
print('Wir sind %r.' % 'am \tbraten')

Wir sind am 	braten.
Wir sind 'am \tbraten'.


Der Operator `%s` konvertiert alles in einen String, inklusive Integers und Floats. Der Operator `%d` konvertiert zuerst Zahlen in Integer, allerdings ohne dabei zu runden. Siehe hier:

In [6]:
print('Ich schrieb %s Programme heute.' % 3.75)
print('Ich schrieb %d Programme heute.' % 3.75)

Ich schrieb 3.75 Programme heute.
Ich schrieb 3 Programme heute.


### Einrückung und Präzision von Gleitkommazahlen

Gleitkommazahlen (Floats) verwenden das Format <code>%5.2f</code>. Hier entspricht <code>5</code> der minimalen Anzahl von Zeichen, die der String enthalten soll; diese werden mit Leerzeichen aufgefüllt, wenn die gesamte Zahl zu wenige Stellen hat. Das folgende <code>.2f</code> steht für die Anzahl der Nachkommastellen. Hier sind ein paar Beispiele:

In [7]:
print('Gleitkommazahl: %5.2f' % (13.144))

Gleitkommazahl: 13.14


In [8]:
print('Gleitkommazahl: %1.0f' % (13.144))

Gleitkommazahl: 13


In [9]:
print('Gleitkommazahl: %1.5f' % (13.144))

Gleitkommazahl: 13.14400


In [10]:
print('Gleitkommazahl: %10.2f' % (13.144))

Gleitkommazahl:      13.14


In [11]:
print('Gleitkommazahl: %25.2f' % (13.144))

Gleitkommazahl:                     13.14


Mehr Informationen zur Stringformatierung mit Gleitkommazahlen findest du auf https://docs.python.org/3/library/stdtypes.html#old-string-formatting

### Mehrfache Formatierung
Nichts verbietet es, mehr als einen Formatierungsoperator einzusetzen:

In [12]:
print('Erster: %s, Zweiter: %5.2f, Dritter: %r' % ('Hi!', 3.1415, 'Ciao!'))

Erster: Hi!, Zweiter:  3.14, Dritter: 'Ciao!'


## Formatierung mit der Methode `.format()`
Ein besser Weg zum Einfügen von Elementen in Print-Statements ist die Stringmethode `.format()`. Die Syntax ist:

    'String hier {} dann auch {}'.format('etwas1','etwas2')
    
Zum Beispiel:

In [13]:
print('Dies ist ein String mit einer {}'.format('Einfügung'))

Dies ist ein String mit einer Einfügung


### Die Methode .format() hat einige Vorteile gegenüber dem Operator %s:

#### 1. Die Einfügung kann per Indizierung erfolgen:

In [14]:
print('Der {2} {1} {0}'.format('Fuchs', 'braune', 'schnelle'))

Der schnelle braune Fuchs


#### 2. Es kann auch mittels Variablen eingefügt werden:

In [15]:
print('Erstes Objekt: {a}, Zweites Objekt: {b}, Drittes Objekt: {c}'.format(
    a=1, b='Zwei', c=12.3))

Erstes Objekt: 1, Zweites Objekt: Zwei, Drittes Objekt: 12.3


#### 3. Elemente können wiederholt verwendet werden:

In [16]:
print('Ein gesparter %s ist ein verdienter %s.' % ('Cent', 'Cent'))
# vs.
print('Ein gesparter {c} ist ein verdienter {c}.'.format(c='Cent'))

Ein gesparter Cent ist ein verdienter Cent.
Ein gesparter Cent ist ein verdienter Cent.


### Ausrichtung, Auffüllen und Präzision mit `.format()`
Mit geschweiften Klammern können Feldlängen, Ausrichtung nach Links/Rechts, Rundungsparameter und mehr festgelegt werden.

In [1]:
print('{0:8} | {1:9}'.format('Frucht', 'Anzahl'))
print('{0:8} | {1:9}'.format('Äpfel', 3.))
print('{0:8} | {1:9}'.format('Orangen', 10))

Frucht   | Anzahl   
Äpfel    |       3.0
Orangen  |        10


Die Voreinstellung von `.format()` ist die Ausrichtung von Text nach links und Zahlen nach rechts. Mit den optionalen Parametern `<`,`^`, oder `>` kann die Ausrichtung nach links, mittig oder rechts definiert werden:

In [2]:
print('{0:<8} | {1:^8} | {2:>8}'.format('Links', 'Mittig', 'Rechts'))
print('{0:<8} | {1:^8} | {2:>8}'.format(11, 22, 33))

Links    |  Mittig  |   Rechts
11       |    22    |       33


Du kannst den Ausrichtungsoperator mit einem Auffülloperator erweitern:

In [3]:
print('{0:=<8} | {1:-^8} | {2:.>8}'.format('Links', 'Mittig', 'Rechts'))
print('{0:=<8} | {1:-^8} | {2:.>8}'.format(11, 22, 33))

Links=== | -Mittig- | ..Rechts
11====== | ---22--- | ......33


Feldlänge und Nachkommazahlen werden ähnlich wie bei der ersten Methode gehandhabt. Die folgenden beiden Print-Statements sind äquivalent:

In [4]:
print('Dies ist meine zweistellige Gleitkommazahl mit zehn Zeichen: %10.2f' % 13.579)
print(
    'Dies ist meine zweistellige Gleitkommazahl mit zehn Zeichen: {0:10.2f}'.format(13.579))

Dies ist meine zweistellige Gleitkommazahl mit zehn Zeichen:      13.58
Dies ist meine zweistellige Gleitkommazahl mit zehn Zeichen:      13.58


Beachte: Auf den Doppelpunkt folgen 5 Leerzeichen und 5 Zeichen für 13.58 insgesamt also 10 Zeichen. 

Mehr Informationen zur Stringmethode `.format()` findest du auf https://docs.python.org/3/library/string.html#formatstrings

## Formatierte Stringliterale (f-strings)

Die in Python 3.6 eingeführten f-Strings haben verschiedene Vorteile gegenüber den älteren, obig beschriebenen Methoden der Stringformatierung. Beispielsweise kannst du normale Variablen direkt im String einbinden, ohne sie als Parameter vorher mit `.format(var)` zu übergeben.

In [5]:
name = 'Fred'

print(f"Er sagte, sein Name sei {name}.")

Er sagte, sein Name sei Fred.


Übergib `!r` für die String-Repräsentation:

In [6]:
print(f"Er sagte, sein Name sei {name!r}")

Er sagte, sein Name sei 'Fred'


#### Gleitkommazahlformatierung mittels `"result: {value:{width}.{precision}}"`

Der Ausdruck `{value:10.4f}` der Methode `.format()` entspricht dem f-String `{value:{10}.{6}}`:

In [7]:
num = 23.45678
print(
    "Dies ist meine vierstellige Gleitkommazahl mit zehn Zeichen: {0:10.4f}".format(num))
print(
    f"Dies ist meine vierstellige Gleitkommazahl mit zehn Zeichen: {num:{10}.{6}}")

Dies ist meine vierstellige Gleitkommazahl mit zehn Zeichen:    23.4568
Dies ist meine vierstellige Gleitkommazahl mit zehn Zeichen:    23.4568


Beachte, dass sich *Präzision* bei f-Strings auf die Gesamtzahl der Stellen der Zahl bezieht, nicht nur auf die Nachkommastellen!
Das passt besser zur wissenschaftlichen Zahlennotation und statistischer Analyse. Unglücklicherweise unterstützen f-Strings nicht das Auffüllen von rechts, selbst wenn Präzision es erlaubt:

In [8]:
num = 23.45
print(
    "Dies ist meine vierstellige Gleitkommazahl mit zehn Zeichen: {0:10.4f}".format(num))
print(
    f"Dies ist meine vierstellige Gleitkommazahl mit zehn Zeichen: {num:{10}.{6}}")

Dies ist meine vierstellige Gleitkommazahl mit zehn Zeichen:    23.4500
Dies ist meine vierstellige Gleitkommazahl mit zehn Zeichen:      23.45


Wenn es erforderlich ist, kannst du immer die `.format()`-Snytax innerhalb eines f-Strings anwenden:

In [9]:
num = 23.45
print(
    "Dies ist meine vierstellige Gleitkommazahl mit zehn Zeichen: {0:10.4f}".format(num))
print(
    f"Dies ist meine vierstellige Gleitkommazahl mit zehn Zeichen: {num:10.4f}")

Dies ist meine vierstellige Gleitkommazahl mit zehn Zeichen:    23.4500
Dies ist meine vierstellige Gleitkommazahl mit zehn Zeichen:    23.4500


Mehr Informationen zu formatierten Stringliteralen findest du auf https://docs.python.org/3/reference/lexical_analysis.html#f-strings

Dies waren die Grundlagen der Stringformatierung.

## Prima!